Move snapping of score into the controller

ChillerDragon opened the pull request:
This commit falls under #7777 as the main motivation is to allow forks of ddnet to change the score value
without having to create a git diff that has to maintained.

I have the same method in my fork already since a while and it allows multiple gamemodes to set the value depending on their needs. Because #7777 was rejected it is not upstreamed to ddnet.

I actually tried to add this very method before in #7598 because it is such a key requirement for custom gametypes. And a common painpoint for modders like me that try to stay close to ddnet upstream.

My interested in proposing this again was caused by the length of this line:

https://github.com/ddnet/ddnet/blob/77c17304eb0a160728b17e932aa215a2b524b66e/src/game/server/player.cpp#L385

And the score being set in 3 different places in the player snap method. So this time I would argue that ddnet it self directly benfits from this cleanup.

## Checklist

- [ ] Tested the change ingame
  - [x] https://github.com/ddnet/ddnet/commit/31f2d6317213977be5f65e4c14f9cea8b669e00a
     - [x] teeworlds 0.7 client
     - [x] ddnet 0.6 client
     - [x] ddnet 0.7 client
  - [ ] [split 0.7 and 0.6 cleaner](https://github.com/ddnet/ddnet/pull/11437#discussion_r2624135730) ([diff to 356298174](https://github.com/ddnet/ddnet/compare/b727d5e9169acbb80eef93b2cc4e3452b831dbc1..3562981747177c538051c029502c93cdd2773bc0))
    - [ ] teeworlds 0.7 client
    - [ ] ddnet 0.6 client
    - [ ] ddnet 0.7 client
  - [x] delete `CPlayer::m_Score` https://github.com/ddnet/ddnet/pull/11437/commits/2562b5de57f3d868309dbe6ef2f317176fa71047
    - [x] teeworlds 0.7 client (looked good until it segfaulted)
    - [x] ddnet 0.6 client
    - [x] ddnet 0.7 client
  - [x] fix conflicts with #11448 and #11442 ([diff](https://github.com/ddnet/ddnet/compare/2562b5de57f3d868309dbe6ef2f317176fa71047..ca5aaf0a90b86ddc0ccbd2d1c6d2f8f5cc734ee6))
    - [x] teeworlds 0.7 client
    - [x] ddnet 0.6 client
    - [x] ddnet 0.7 client
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)
- [x] I didn't use generative AI to generate more than single-line completions

ChillerDragon commented on src/game/server/player.cpp:
```diff
@@ -332,25 +332,7 @@ void CPlayer::Snap(int SnappingClient)
-	{
-		Score = FinishTime::NOT_FINISHED_TIMESCORE;
-	}
+	int Score = GameServer()->m_pController ? GameServer()->m_pController->SnapPlayerScore(SnappingClient, this) : 0;
```
There shouldn't be any snapshots while there is no controller.

CC #11296

I just didn't want to be the guy who wrote the segfaulting line if something changes and this happens. But yolo.

heinrich5991 commented on src/game/server/gamemodes/DDRace.cpp:
```diff
@@ -118,6 +120,31 @@ void CGameControllerDDRace::SetArmorProgress(CCharacter *pCharacter, int Progres
+	// seconds to finish the map.
+	if(-pPlayer->m_Score.value() == FinishTime::NOT_FINISHED_TIMESCORE)
+		return -pPlayer->m_Score.value() - 1;
+	return -pPlayer->m_Score.value();
```
I think it'd be good if this delineated 0.6/0.7 code more cleanly. I.e. have one top-level `if(Server()->IsSixup(SnappingClient)` and not handle both throughout the function.

heinrich5991 reviewed (commented):
Thanks. Looks like a useful cleanup for DDNet as well.

ChillerDragon commented on src/game/server/gamemodes/DDRace.cpp:
```diff
@@ -118,6 +120,31 @@ void CGameControllerDDRace::SetArmorProgress(CCharacter *pCharacter, int Progres
+
+	// shift the time by a second if the player actually took 9999
+	// seconds to finish the map.
+	if(-pPlayer->m_Score.value() == FinishTime::NOT_FINISHED_TIMESCORE)
```
I see thanks for clarifying this makes sense now.

So you mean that we only use this one

https://github.com/ddnet/ddnet/blob/7f43c1194f6e647fb0cd71e19802e5843095a435/src/game/server/scoreworker.h#L257

And delete this

https://github.com/ddnet/ddnet/blob/7f43c1194f6e647fb0cd71e19802e5843095a435/src/game/server/player.h#L107

and also delete this

https://github.com/ddnet/ddnet/blob/7f43c1194f6e647fb0cd71e19802e5843095a435/src/engine/server/server.h#L164

One is a float the other two are optional ints and they are passed around a bit. I am not keen on doing this bigger change. Can someone else pick this up in a follow up pr? I don't want to refactor the entire code and create new bugs.

ChillerDragon commented:
Yikes so after finishing a second time with the teeworlds client at https://github.com/teeworlds/teeworlds/commit/5d682733e482950f686663c129adc4b751c8d790

https://github.com/teeworlds/teeworlds/blob/5d682733e482950f686663c129adc4b751c8d790/src/engine/shared/snapshot.cpp#L388

my `mem_copy` destination `pNewData` is a `nullptr`.

```
[2025-12-17 14:38:15][chat_commands]: adding server chat command: name='whispers' args='?i['0'|'1']' help='Toggle receiving whispers'
[2025-12-17 14:38:15][client]: dropped weird message 'invalid' (0), failed on '(type out of range)'
[2025-12-17 14:38:16][datafile]: loading data index=7 size=62 uncompressed=152
[2025-12-17 14:38:16][client]: dropped weird message 'invalid' (0), failed on '(type out of range)'
[2025-12-17 14:38:24][chat/server]: *** 'teeworlds' has left the game (Too weak connection (not acked for 100 seconds))
[2025-12-17 14:38:29][race]:  2: (1)teeworlds: finished in 00:06.50
[2025-12-17 14:38:29][chat/server]: *** '(1)teeworlds' has set a new personal record: 00:06.50 (10.00 seconds faster)

Thread 1 "teeworlds" received signal SIGSEGV, Segmentation fault.
366             movq    %rsi, (%rdi)
(gdb) bt
#0  __memcpy_avx512_unaligned_erms () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:366
#1  0x00005555556db844 in mem_copy (dest=0x0, source=0x7ffffffdc93c, size=12) at /home/chiller/Desktop/gaming/teeworlds/src/base/system.c:340
#2  0x00005555556d2c02 in CSnapshotDelta::UnpackDelta (this=0x7ffff6bbd8d0, pFrom=0x5555574af8c0, pTo=0x7ffffffec6f0, pSrcData=0x7ffffffdc6f0, DataSize=600)
    at /home/chiller/Desktop/gaming/teeworlds/src/engine/shared/snapshot.cpp:388
#3  0x0000555555575da0 in CClient::ProcessServerPacket (this=0x7ffff6926010, pPacket=0x7fffffffc740) at /home/chiller/Desktop/gaming/teeworlds/src/engine/client/client.cpp:1452
#4  0x0000555555576740 in CClient::PumpNetwork (this=0x7ffff6926010) at /home/chiller/Desktop/gaming/teeworlds/src/engine/client/client.cpp:1586
#5  0x0000555555577168 in CClient::Update (this=0x7ffff6926010) at /home/chiller/Desktop/gaming/teeworlds/src/engine/client/client.cpp:1751
#6  0x00005555555789d0 in CClient::Run (this=0x7ffff6926010) at /home/chiller/Desktop/gaming/teeworlds/src/engine/client/client.cpp:2095
#7  0x000055555557b36d in main (argc=2, argv=0x7fffffffcc78) at /home/chiller/Desktop/gaming/teeworlds/src/engine/client/client.cpp:2702
```

ChillerDragon commented:
Okay I finished updating the code and testing it. There are no more known bugs or todos from my side.
Please review 2562b5de57f3d868309dbe6ef2f317176fa71047 closely I am sure it broke something big time.

(ideally we would merge without it :p but eh)

ChillerDragon commented on src/game/server/teams.cpp:
```diff
@@ -859,9 +859,10 @@ void CGameTeams::OnFinish(CPlayer *pPlayer, int TimeTicks, const char *pTimestam
+	if(!Score.has_value() || TTime < Score.value())
 	{
-		pPlayer->m_Score = TTime;
+		Server()->SetClientScore(ClientId, TTime);
```
@heinrich5991 I placed it here

CC https://github.com/ddnet/ddnet/pull/11437/changes#r2627496034

ChillerDragon commented on src/game/server/player.cpp:
```diff
@@ -182,8 +181,6 @@ void CPlayer::Tick()
 	if(m_ChatScore > 0)
 		m_ChatScore--;
 
-	Server()->SetClientScore(m_ClientId, m_Score);
```
https://github.com/ddnet/ddnet/pull/11437/changes#r2634873598
